internal/cri/server/images/image_pull.go

```diff
@@ -479,7 +478,9 @@ func (c *CRIImageService) registryHosts(ctx context.Context, credentials func(ho
 
 			authorizer := docker.NewDockerAuthorizer(
 				docker.WithAuthClient(client),
-				docker.WithAuthCreds(credentials))
+				docker.WithAuthCreds(credentials),
+				docker.WithAuthHeader(c.config.Registry.Headers),
```

Comment:
is there per-host overrides possible as well? Sorry I do not know this code very well

Reply:
Unfortunately the header override is not per-host in the old cri registry config.

https://github.com/containerd/containerd/blob/2c042ae0de7b9d69df8df8e370e2d802f10f9ee5/internal/cri/config/config.go#L236-L237

The old registry config in cri looks relatively limited (from the code around). Only the following are per host:
1. Mirrors: host -> list of mirror endpoint strings (no other per mirror config like headers).
2. Configs: host -> RegistryConfig which only has a AuthConfig
3. Auths: host -> AuthConfig.

Reply:
I think there is a good argument here to not fix this and rather point to the hosts configurations as the fix. This may be a fix but could also count as a behavior change in a deprecated feature. @SergeyKanzhelev is that a concern you have about it not being per-host?

Reply:
This SGTM.

> also count as a behavior change in a deprecated feature

This cri global `Headers` doesn't seem to be deprecated 🤔 Or we should mark it as deprecated as well, same as Auths/Configs/etc?

https://github.com/containerd/containerd/blob/23500b8015c6f5c624ec630fd1377a990e9eccfb/internal/cri/config/config.go#L234-L237